src/common/authenticator.cpp

```diff
@@ -122,64 +133,230 @@ int Authenticator::HMacSha256(const void* key, int key_size,
+    if (token.encticket().empty() || token.encclientidentity().empty()) {
+        LOG(ERROR) << "token info is not enough";
+        return false;
     }
-    *q = 0;
-    if ((safe % 3) == 1) {
-        *(q - 1) = '=';
-        *(q - 2) = '=';
+    // decrypt ticket
+    Ticket ticket;
+    bool needRedecByLastKey = false;
+    unsigned char uiv[AES_BLOCK_SIZE];
+    std::memcpy(uiv, iv_, AES_BLOCK_SIZE);
+    std::string ticketStr;
+    int ret = Encryptor::AESDecrypt(options_.key, uiv, token.encticket(),
+        &ticketStr);
+    if (ret != 0) {
+        LOG(WARNING) << "aes decrypt ticket failed when VerifyCredential"
+                     << "used current key, then will try last key: "
+                     << options_.lastKey;
+        needRedecByLastKey = true;
+    } else {
+        if (!ticket.ParseFromString(ticketStr)) {
+            LOG(WARNING) << "aes decrypt ticket failed when VerifyCredential"
+                         << "used current key, then will try last key: "
+                            << options_.lastKey;
+            needRedecByLastKey = true;
+        }
+    }
+    if (needRedecByLastKey) {
+        // maybe in the key rotation phase
+        if (options_.lastKey.size() == 0 || LastKeyExpired()) {
+            LOG(ERROR) << "aes decrypt ticket failed with current key and last"
+                       << " key is empty or expiration when VerifyCredential"
+                       << ", lastKey size = " << options_.lastKey.size()
+                       << ", lastKey expiration = " << lastKeyExpiration_;
+            return false;
+        } else {
+            std::memcpy(uiv, iv_, AES_BLOCK_SIZE);
+            ret = Encryptor::AESDecrypt(options_.lastKey, uiv,
+                token.encticket(), &ticketStr);
+            if (ret != 0) {
+                LOG(ERROR) << "aes decrypt ticket failed with last key"
+                           << "when VerifyCredential";
+                return false;
+            }
+        }
+        if (!ticket.ParseFromString(ticketStr)) {
+            LOG(ERROR) << "aes decrypt ticket failed with last key"
+                       << " when VerifyCredential";
+            return false;
+        }
+        LOG(INFO) << "verify success with last key, lastKey expiration = "
+                  << lastKeyExpiration_;
     }
-    if ((safe % 3) == 2)
-        *(q - 1) = '=';
 
-    if (pp != src)
-        free(pp);
+    // check timeout
+    auto now = TimeUtility::GetTimeofDaySec();
+    if (ticket.expiration() < now) {
+        LOG(ERROR) << "ticket is expired, now = " << now
+                   << ", expiration = " << ticket.expiration();
+        return false;
+    }
+
+    // decrypt client identity
+    std::memcpy(uiv, iv_, AES_BLOCK_SIZE);
+    std::string cInfoString;
+    ret = curve::common::Encryptor::AESDecrypt(ticket.sessionkey(),
+        uiv, token.encclientidentity(), &cInfoString);
+    if (ret != 0) {
+        LOG(ERROR) << "aes decrypt cInfo failed when VerifyCredential";
```

Comment:
ditto, if client send an invalid request to chunkserver, chunkserver shouldn't output an error log.

Reply:
ok, change to warnning log, The purpose of keeping these logs is to know the specific reason when authentication fails.